app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,544 @@
+          Log.v(TAG, "Canceled");
+          break;
+        default:
+          break;
+      }
+    }
+  }
+
+  @Override protected void onDestroy() {
+    super.onDestroy();
+    presenter.detachView();
+  }
+
+  private void setUpToolbar() {
+    Toolbar toolbar = findViewById(R.id.toolbar);
+    setSupportActionBar(toolbar);
+    getSupportActionBar().setTitle(getString(R.string.menu_host_books));
+    getSupportActionBar().setHomeButtonEnabled(true);
+    getSupportActionBar().setDisplayHomeAsUpEnabled(true);
+
+    toolbar.setNavigationOnClickListener(v -> onBackPressed());
+  }
+
+  private void setupLocationServices() {
+    LocationRequest locationRequest = new LocationRequest();
+    locationRequest.setInterval(10);
+    locationRequest.setSmallestDisplacement(10);
+    locationRequest.setFastestInterval(10);
+    locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
+    LocationSettingsRequest.Builder builder = new
+        LocationSettingsRequest.Builder();
+    builder.addLocationRequest(locationRequest);
+
+    task = LocationServices.getSettingsClient(this).checkLocationSettings(builder.build());
+
+    locationSettingsResponseBuilder();
+  }
+
+  private void locationSettingsResponseBuilder() {
+    task.addOnCompleteListener(task -> {
+      try {
+        LocationSettingsResponse response = task.getResult(ApiException.class);
+        // All location settings are satisfied. The client can initialize location
+        // requests here.
+
+        startService(ACTION_TURN_ON_AFTER_O);
+
+        //}
+      } catch (ApiException exception) {
+        switch (exception.getStatusCode()) {
+          case LocationSettingsStatusCodes.RESOLUTION_REQUIRED:
+            // Location settings are not satisfied. But could be fixed by showing the
+            // user a dialog.
+            try {
+              // Cast to a resolvable exception.
+              ResolvableApiException resolvable = (ResolvableApiException) exception;
+              // Show the dialog by calling startResolutionForResult(),
+              // and check the result in onActivityResult().
+              resolvable.startResolutionForResult(
+                  ZimHostActivity.this,
+                  LOCATION_SETTINGS_PERMISSION_RESULT);
+            } catch (IntentSender.SendIntentException e) {
+              // Ignore the error.
+            } catch (ClassCastException e) {
+              // Ignore, should be an impossible error.
+            }
+            break;
+          case LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE:
+            // Location settings are not satisfied. However, we have no way to fix the
+            // settings so we won't show the dialog.
+            break;
+          default:
+            break;
+        }
+      }
+    });
+  }
+
+  //Advice user to turn on hotspot manually for API<26
+  private void startHotspotDialog() {
+    AlertDialog.Builder builder = new AlertDialog.Builder(this, dialogStyle());
```

Comment:
if you are using a builder then you should use it like a builder and 
```
builder.chainMethodCalls()
.like()
.this()
.and()
.end()
.with()
.build()
```

Reply:
Replacing this with `AlertDialogShower`